syntax = "proto3";


service ChainReplicationService {
  rpc IsAliveCheck (Empty) returns (IsAliveResponse) {}
  rpc GetDataStores (Empty) returns (DataStoreList){}
  rpc TransferChain (Chain) returns (ChainAcceptedResponse){}
  rpc WriteData (WriteRequest) returns (Empty){}
  rpc SendData (NodeId) returns (DataResponse) {}
}

message Empty {}


message IsAliveResponse {
  bool alive = 1;
}


message Chain {
  repeated string chain = 1;
  string head = 2;
  string tail = 3;
}

message ChainAcceptedResponse {
  bool accepted = 1;
}

message DataStoreList {
  repeated string data_store_ids = 1;
}

message WriteRequest {
  string book = 1;
  float price = 2;
  string id = 3;
}

message DataResponse {
  repeated string books = 1;
  repeated float prices = 2;
}

message NodeId {
  string store_id = 1;
  int64 node_id = 2;
}


